include/dali/core/per_stream_pool.h

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_PER_STREAM_POOL_H_
+#define DALI_CORE_PER_STREAM_POOL_H_
+
+#include <cuda_runtime.h>
+#include <memory>
+#include <vector>
+#include <utility>
+#include <mutex>
+#include "dali/core/cuda_event.h"
+
+namespace dali {
+
+/**
+ * @brief Represents a pool of per-device objects which can be leased to perform job on a stream
+ *
+ * The PerStreamPool class manages objects which can be leased for the purpose of fulfilling some
+ * task on a particular device and stream.
+ * When a request is made, the caller must specify a CUDA stream (and optionally device, otherwise
+ * current device is used) for the task associated with the object. The pool returns a "lease"
+ * RAII wrapper. When the wrapper goes out of scope, the object is returned to the pool
+ * and will be available for lease again when the stream finishes the work that's been scheduled
+ * until the lease ended.
+ * Additionally, if `reuse_pending_on_same_stream` flag is set, subsequent requests for an object
+ * specifying the same stream will return that object even if the associated work has not yet
+ * completed - this is useful for objects which are safe for immediate reuse on the same stream,
+ * e.g. device buffers.
+ * If there are no available objects for current device, a new object is created and leased
+ * immediately - it will be returned to the pool when the lease is over.
+ *
```

Comment:
I'd add some description about PerDevicePool as well. It is not clear to me why it is call per-device, since it is also per stream with the added fact that we don't share the object when there is pending work in the object associated with that stream

Reply:
Actually, if this flag is disabled, it's no longer PerStream - the stream parameter is used solely for scheduling the completion events.